f5 vxlan integration with sdn

rajatchopra opened the pull request:
This PR implements the VxLAN integration of F5 with a vxlan based overlay. Three pieces of work primarily:

1. Router plugin now watches for Nodes in the cluster (to program the F5's vxlan FDB; in the case of template router its a noop)
2. F5 router programs the VxLAN devices/routing on the F5 BIG IP on initialization
3. Command line changes to accept the special parameters needed (including a boolean that disables the setup)

rajatchopra commented:
[test]

knobunc commented:
[test]

chen23 commented:
FYI in the policy section of the code the use of the Legacy flag will only work in TMOS version 12.1 or newer.  Ideally have a version check to only add the flag if TMOS >= 12.1 

```
policyPayload := f5Policy{
		Name:     policyName,
		Controls: []string{"forwarding"},
		Requires: []string{"http"},
		Strategy: "best-match",
		Legacy:   true,
}
```
Python example of version check here: https://github.com/f5devcentral/f5-icontrol-codeshare-python/blob/master/kubernetes-example/k8s2f5.py#L328

chen23 commented:
The auth code looks to only support basicauth.  This will limit authentication to local accounts and not any remote (radius/tacacs), etc...  https://github.com/rajatchopra/origin/blob/3ad39a5c60f344d8fa6cb673f021effe9e139c4b/pkg/router/f5/f5.go#L413 

```
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")
	req.SetBasicAuth(f5.username, f5.password)

client := &http.Client{Transport: tr}
```
Python example: https://github.com/F5Networks/f5-icontrol-rest-python/blob/5c43b0f6424a5f495c50fd2a318e172d5c3d7e37/icontrol/authtoken.py 
DevCentral article: https://devcentral.f5.com/articles/demystifying-icontrol-rest-part-6-token-based-authentication

chen23 commented:
Doing a quick scan of the code I'm not seeing an explicit save of the running config after it is generated.  https://github.com/F5Networks/f5-common-python/issues/342 

iControl REST sits in front of tmsh

https://support.f5.com/kb/en-us/products/big-ip_ltm/manuals/product/bigip-system-essentials-11-6-0/2.html

_"When you use the Traffic Management Shell (tmsh) to configure the system, you must explicitly issue a save command to store the configuration data that you have generated. Otherwise, the newly-generated configuration data is not actually stored on the system. For more information about tmsh, see the Traffic Management Shell (tmsh) Reference Guide."_

kmunson1973 commented:
You can save the running config by issuing a POST /mgmt/tm/sys/config with the following json payload.
 {
 "command":"save"
 }

knobunc commented on pkg/router/f5/f5.go:
```diff
@@ -828,11 +911,77 @@ func (f5 *f5LTM) Initialize() error {
+func checkIPAndGetMac(ipStr string) (string, error) {
+	ip := net.ParseIP(ipStr)
+	if ip == nil {
+		errStr := fmt.Sprintf("'%s' is not a valid IP address for a vtep", ipStr)
```
This implies that the vtep has other restrictions.  Perhaps:
  'vtep IP '%s' is not a valid IP address'

knobunc commented on pkg/router/f5/f5.go:
```diff
@@ -828,11 +911,77 @@ func (f5 *f5LTM) Initialize() error {
 	return nil
 }
 
+func checkIPAndGetMac(ipStr string) (string, error) {
```
Should this have vtep in it?  Or is the package scope enough to indicate that.

rajatchopra replied:
Its just a util function. The package scope is enough I guess.

marun replied:
Consider changing this function name to be more descriptive (e.g. convertIpToVtepMac).

marun replied:
Also, consider adding a unit test.

knobunc commented:
Repo problems re-[test]

knobunc commented:
@smarterclayton PTAL.  The tests are still being worked on, but can you please look over the code to see if it is good?

marun commented on pkg/router/f5/f5.go:
```diff
@@ -461,6 +481,68 @@ func (f5 *f5LTM) delete(url string, result interface{}) error {
+	}
+	// create the net self IP
+	err = f5.post(selfUrl, netSelfPayload, nil)
+	if err != nil && err.(F5Error).httpStatusCode != 409 {
```
ditto re: named constant

marun commented on pkg/router/f5/plugin.go:
```diff
@@ -466,10 +485,46 @@ func (p *F5Plugin) deleteRoute(routename string) error {
+			glog.Warningf("Error in obtaining IP address of newly added node %s - %v", node.Name, err)
+			return nil
+		}
+		p.F5Client.AddVtep(ip)
```
Should this call be returning the error that is potentially returned by AddVtep()?

marun commented on pkg/router/f5/plugin.go:
```diff
@@ -466,10 +485,46 @@ func (p *F5Plugin) deleteRoute(routename string) error {
+			glog.Warningf("Error in obtaining IP address of deleted node %s - %v", node.Name, err)
+			return nil
+		}
+		p.F5Client.RemoveVtep(ip)
```
Should this call be returning the error that is potentially returned by RemoveVtep?

marun commented on pkg/router/f5/plugin.go:
```diff
@@ -466,10 +485,46 @@ func (p *F5Plugin) deleteRoute(routename string) error {
+		}
+		p.F5Client.RemoveVtep(ip)
+	case watch.Modified:
+		// ignore the modified event. Change in IP address of the node is not supported.
```
What's the implication of not supporting node address changes?  Is it just going to be documented?  Should there be logging to diagnose issues if/when address changes occur?

ramr replied:
I had the same question re: changes and asked Rajat on this. He said on the openshift-sdn side we don't support IP address changes for a node.

ramr commented on pkg/cmd/infra/router/f5.go:
```diff
@@ -88,6 +103,9 @@ func (o *F5Router) Bind(flag *pflag.FlagSet) {
 	flag.StringVar(&o.PartitionPath, "f5-partition-path", util.Env("ROUTER_EXTERNAL_HOST_PARTITION_PATH", f5plugin.F5DefaultPartitionPath), "The F5 BIG-IP partition path to use")
+	flag.StringVar(&o.InternalAddress, "f5-internal-address", util.Env("ROUTER_EXTERNAL_HOST_INTERNAL_ADDRESS", ""), "The F5 BIG-IP internal interface's IP address")
+	flag.StringVar(&o.VxlanGateway, "f5-vxlan-gateway-cidr", util.Env("ROUTER_EXTERNAL_HOST_VXLAN_GW_CIDR", ""), "The F5 BIG-IP gateway-ip-address/cidr-mask for setting up the VxLAN")
+	flag.BoolVar(&o.SetupOSDNVxLAN, "f5-setup-osdn-vxlan", false, "Flag to enable VxLAN setup for integration openshif-sdn with the F5 BIG-IP")
```
so this needs to be set with a command line arg option?  It maybe better to implicitly set it if `f5-internal-address` and `f5-vxlan-gateway-cidr` is set - basically one without the other(s) doesn't make sense does it?

ramr commented on pkg/cmd/infra/router/f5.go:
```diff
@@ -108,6 +126,10 @@ func (o *F5Router) Validate() error {
 		return errors.New("F5 HTTP and HTTPS vservers cannot both be blank")
 	}
 
+	if o.SetupOSDNVxLAN && (len(o.VxlanGateway) == 0 || len(o.InternalAddress) == 0) {
```
See above comment re: `o.SetupOSDNVxLAN`.

ramr commented on pkg/router/controller/controller.go:
```diff
@@ -51,6 +54,8 @@ func (c *RouterController) Run() {
 	}
 	go utilwait.Forever(c.HandleRoute, 0)
 	go utilwait.Forever(c.HandleEndpoints, 0)
+	c.HandleNode()
```
Is this needed? The one from the goroutine ahead should be called soon anyway.

ramr commented on pkg/router/controller/factory/factory.go:
```diff
@@ -35,10 +36,11 @@ type RouterControllerFactory struct {
 	return &RouterControllerFactory{
 		KClient:        kc,
 		OSClient:       oc,
+		NClient:        kc,
```
Yeah, if its same as kc - we should cleanup to just use 1 `?Client`. Maybe cleanup in follow on work. I know changing it would touch a lot more code in this PR.

ramr replied:
A comment/todo here or in the type definition would be useful.

ramr commented on pkg/router/controller/unique_host.go:
```diff
@@ -84,6 +84,11 @@ func (p *UniqueHost) HandleEndpoints(eventType watch.EventType, endpoints *kapi.
+func (p *UniqueHost) HandleNode(eventType watch.EventType, node *kapi.Node) error {
+	return p.plugin.HandleNode(eventType, node)
+}
+
```
FYI, depending on which PR merges first - https://github.com/openshift/origin/pull/11201 - will need to add HandleNode to the `host_admitter` plugin.

ramr commented on pkg/router/f5/f5.go:
```diff
@@ -461,6 +481,68 @@ func (f5 *f5LTM) delete(url string, result interface{}) error {
+	// create the profile
+	url := fmt.Sprintf("https://%s/mgmt/tm/net/tunnels/vxlan", f5.host)
+	profilePayload := f5CreateVxLANProfilePayload{
+		Name:         "vxlan-ose",
```
Can we use a constant for `vxlan-ose`? Might be a candidate for a parameter name, maybe?

ramr commented on pkg/router/f5/f5.go:
```diff
@@ -461,6 +481,68 @@ func (f5 *f5LTM) delete(url string, result interface{}) error {
+	// create the tunnel
+	url = fmt.Sprintf("https://%s/mgmt/tm/net/tunnels/tunnel", f5.host)
+	tunnelPayload := f5CreateVxLANTunnelPayload{
+		Name:         "vxlan5000",
```
Same as above. Can we use a constant for `vxlan5000`? Might be a candidate for a parameter name, maybe?

ramr commented on pkg/router/f5/f5.go:
```diff
@@ -461,6 +481,68 @@ func (f5 *f5LTM) delete(url string, result interface{}) error {
+		InheritedTrafficGroup: "false",
+		TrafficGroup:          path.Join(f5.partitionPath, "traffic-group-local-only"),
+		Unit:                  0,
+		Vlan:                  path.Join(f5.partitionPath, "vxlan5000"),
```
ditto re: named constant!! Copy-pasta from @marun comments!! ;^)

ramr commented on pkg/router/f5/f5.go:
```diff
@@ -828,11 +912,77 @@ func (f5 *f5LTM) Initialize() error {
+		return err
+	}
+
+	url := fmt.Sprintf("https://%s/mgmt/tm/net/fdb/tunnel/%s~vxlan5000/records", f5.host, strings.Replace(f5.partitionPath, "/", "~", -1))
```
And same comment here re: using a constant for `vxlan5000`.

ramr commented on pkg/router/f5/f5.go:
```diff
@@ -828,11 +912,77 @@ func (f5 *f5LTM) Initialize() error {
+		return err
+	}
+
+	url := fmt.Sprintf("https://%s/mgmt/tm/net/fdb/tunnel/%s~vxlan5000/records/%s", f5.host, strings.Replace(f5.partitionPath, "/", "~", -1), macAddr)
```
And same comment here re: using a constant for `vxlan5000`.

ramr commented on pkg/router/f5/types.go:
```diff
@@ -120,6 +123,8 @@ type f5Policy struct {
 	// other possible values are all-match and first-match.
 	Strategy string `json:"strategy"`
+
+	Legacy bool `json:"legacy"`
```
I was curious on this one - thanks for explaining it. So if we ever decide to lock down to a specific version of the `F5` api, then we would have to remove this - maybe adding a comment here would be helpful. Thx

smarterclayton replied:
Godoc

ramr commented on pkg/router/f5/types.go:
```diff
@@ -106,6 +106,9 @@ type f5Policy struct {
 	Name string `json:"name"`
 
+	// Partition name for the policy
+	TmPartition string `json:"tmPartition"`
```
Felt a little weird that this one is called `TmPartition` but every other F5 api call payload uses `Partition`. Thanks for explaining that one.

smarterclayton replied:
Update the godoc explaining what TmPartition means or be more explicit

ramr commented on pkg/router/f5/f5.go:
```diff
@@ -461,6 +481,68 @@ func (f5 *f5LTM) delete(url string, result interface{}) error {
+		LocalAddress: f5.internalAddress,
+		Mode:         "bidirectional",
+		Mtu:          "0",
+		Profile:      path.Join(f5.partitionPath, "vxlan-ose"),
```
constant for `vxlan-ose`.

ramr commented on pkg/router/controller/factory/factory.go:
```diff
@@ -65,6 +67,13 @@ func (factory *RouterControllerFactory) Create(plugin router.Plugin) *controller
+	nodeEventQueue := oscache.NewEventQueue(cache.MetaNamespaceKeyFunc)
+	cache.NewReflector(&nodeLW{
+		client: factory.NClient,
+		field:  factory.Fields,
```
Dunno - would it work if you only have a subset of nodes added to the F5 config and the endpoints for a service are on nodes that are not part of the set?

rajatchopra replied:
I see no harm in keeping the fields/filtering. Possible use case with FDB being populated for only a portion of the cluster (say a datacenter).

smarterclayton replied:
It does not make sense to just apply field labeling to nodes - we should never apply the same field selector to two types.

ramr commented:
LGTM

marun commented:
networking job flaked with #11452 re-[test]

eparis commented:
[merge] (in like 2 weeks because the queue is so backed up)

smarterclayton commented on pkg/router/controller/controller.go:
```diff
@@ -51,6 +52,7 @@ func (c *RouterController) Run() {
 	}
 	go utilwait.Forever(c.HandleRoute, 0)
 	go utilwait.Forever(c.HandleEndpoints, 0)
+	go utilwait.Forever(c.HandleNode, 0)
```
Why is this always on if it's only used in F5?

smarterclayton commented on pkg/router/controller/factory/factory.go:
```diff
@@ -27,6 +27,7 @@ import (
 type RouterControllerFactory struct {
 	KClient        kclient.EndpointsNamespacer
 	OSClient       osclient.RoutesNamespacer
+	NClient        kclient.NodesInterface
```
NamespaceClient, abbreviations were a special case but shouldn't be propogared.

smarterclayton commented:
Removing from merge queue until my comments are addressed.

rajatchopra commented:
@smarterclayton I have addressed all your feedback. Primarily two:
1. Node watch loop does not run unless its F5 && a vxlan setup is needed
2. All fields/labels are put in for node watch (we have no use case yet where only a portion of the sdn needs to be addressed by F5)

Marking it [merge]. Please feel free to revert the commit if anything is loose.

TODO: 
Document on how to use the native VxLAN feature

eparis commented:
[merge] you stoopid flake. Merge!

eparis commented:
I'm an idiot, it wasn't a flake!
```
test/integration/router_stress_test.go:285: not enough arguments in call to factory.Create
```

eparis commented:
back in at #5. wait and see!

deads2k commented on test/testdata/bootstrappolicy/bootstrap_cluster_roles.yaml:
```diff
@@ -1776,6 +1776,14 @@ items:
     - ""
     attributeRestrictions: null
     resources:
+    - nodes
```
@smarterclayton didn't you want this role bindable by project-admins?  This will prevent that from being possible.

As for our system router, it doesn't worry overly much, but for per-project routers, this makes it impossible to create those.

eparis replied:
Its also required. If we expect the F5 to participate in the SDN it needs to know about the nodes its talking. We don't have OpenShift-sdn doing that. If this is a big problem it would mean splitting the f5 into its own role, I guess?

deads2k replied:
> Its also required. If we expect the F5 to participate in the SDN it needs to know about the nodes its talking. We don't have OpenShift-sdn doing that. If this is a big problem it would mean splitting the f5 into its own role, I guess?

Or adding it as the delta required.

smarterclayton commented:
As David noted this causes a security vulnerability because it's escalating the privileges granted to system:router.  I'm going to push a change that removes the node permission granted here and to fix the issue you can open another one (instead of reverting here).

smarterclayton commented:
Opened a PR that removes the permission, we can iterate on the next solution at a lower pace.

smarterclayton commented:
It's not a flake.  The tests don't compile.

smarterclayton commented:
We could also create system:local-router or system:personal-router.

On Mon, Oct 24, 2016 at 9:14 AM, David Eads <notifications@github.com>
wrote:

> *@deads2k* commented on this pull request.
> ------------------------------
>
> In test/testdata/bootstrappolicy/bootstrap_cluster_roles.yaml
> <https://github.com/openshift/origin/pull/11181>:
>
> > @@ -1776,6 +1776,14 @@ items:
>      - ""
>      attributeRestrictions: null
>      resources:
> +    - nodes
>
> Its also required. If we expect the F5 to participate in the SDN it needs
> to know about the nodes its talking. We don't have OpenShift-sdn doing
> that. If this is a big problem it would mean splitting the f5 into its own
> role, I guess?
>
> Or adding it as the delta required.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/11181>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p5OJReoyNdhUMz449EXSxXww4BvEks5q3K84gaJpZM4KLn0M>
> .
>

eparis commented:
You mean why I said `I'm an idiot, it wasn't a flake!`

smarterclayton commented:
Integration tests aren't compiling

On Oct 23, 2016, at 4:45 PM, OpenShift Bot <notifications@github.com> wrote:

continuous-integration/openshift-jenkins/merge FAILURE (
https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/10508/) (Base
Commit: 35b539d
<https://github.com/openshift/origin/commit/35b539d8aa39e0e79296b6d3fd8aaf21ea8bce63>
)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/11181#issuecomment-255613189>,
or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p2r1fYHWcAjiUQ-TekgW_5ppbsHqks5q28dfgaJpZM4KLn0M>
.

smarterclayton commented:
The least disruptive thing is to create a new role with the additional
permission and require admins to set it up.  We can't retroactively add it
to old clusters (reconcile) if anyone has used it, and it's a very big and
surprising role.

So create system:router-.... (david i need a name suggestion) that contains
the node role.

On Mon, Oct 24, 2016 at 10:11 AM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> We could also create system:local-router or system:personal-router.
>
> On Mon, Oct 24, 2016 at 9:14 AM, David Eads <notifications@github.com>
> wrote:
>
>> *@deads2k* commented on this pull request.
>> ------------------------------
>>
>> In test/testdata/bootstrappolicy/bootstrap_cluster_roles.yaml
>> <https://github.com/openshift/origin/pull/11181>:
>>
>> > @@ -1776,6 +1776,14 @@ items:
>>      - ""
>>      attributeRestrictions: null
>>      resources:
>> +    - nodes
>>
>> Its also required. If we expect the F5 to participate in the SDN it needs
>> to know about the nodes its talking. We don't have OpenShift-sdn doing
>> that. If this is a big problem it would mean splitting the f5 into its own
>> role, I guess?
>>
>> Or adding it as the delta required.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/openshift/origin/pull/11181>, or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABG_p5OJReoyNdhUMz449EXSxXww4BvEks5q3K84gaJpZM4KLn0M>
>> .
>>
>
>

smarterclayton commented:
GitHub is delivering emails I sent on Friday.  Thanks github.

deads2k commented:
`system:rouer-privileged`, `system:rouer-extensions`?
